### ISSUE TITLE
Application controller excessive reconciling

### ISSUE THREAD

**Issue by paveq** [argoproj/argo-cd#9014]
**Describe the bug**

When Kubernetes resource contained within Argo-CD application changes rapidly in target cluster and namespace (due to bug in eg. some other operator), it causes application controller getting loaded.

**To Reproduce**

Have a resource updating rapidly in application's target namespace.

**Expected behavior**

Application controller should detect this happens, and backoff refreshing applications exponentially, potentially remove watches for specific time on that application.

Even if root cause of resource updating rapidly is certainly bug somewhere else, it should not cause argo-cd to fail or spend excess amount of CPU to constantly reconcile application.

**Version**
```
2.3.3
```

**Logs**

In case here, we have argo-events sensor object updating rapidly due to unknown cause, but we've seen this happen with other CR's and operators too.

```
time="2022-04-06T10:44:45Z" level=info msg="Refreshing app status (controller refresh requested), level (1)" application=argo-events-ci
time="2022-04-06T10:44:45Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: argo-events)" application=argo-events-ci
time="2022-04-06T10:44:45Z" level=info msg="getRepoObjs stats" application=argo-events-ci build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=5 unmarshal_ms=5 version_ms=0
time="2022-04-06T10:44:45Z" level=info msg="Skipping auto-sync: most recent sync already to 453b2e55216b2de40fd95d008f96f24a654a3691" application=argo-events-ci
time="2022-04-06T10:44:45Z" level=info msg="No status changes. Skipping patch" application=argo-events-ci
time="2022-04-06T10:44:45Z" level=info msg="Reconciliation completed" application=argo-events-ci dedup_ms=0 dest-name= dest-namespace=argo-events dest-server="https://kubernetes.default.svc" diff_ms=2 fields.level=1 git_ms=6 health_ms=0 live_ms=1 settings_ms=0 sync_ms=0 time_ms=19
time="2022-04-06T10:44:46Z" level=info msg="Refreshing app status (controller refresh requested), level (1)" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: argo-events)" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="getRepoObjs stats" application=argo-events-ci build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=6 unmarshal_ms=6 version_ms=0
time="2022-04-06T10:44:46Z" level=info msg="Skipping auto-sync: most recent sync already to 453b2e55216b2de40fd95d008f96f24a654a3691" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="No status changes. Skipping patch" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="Reconciliation completed" application=argo-events-ci dedup_ms=0 dest-name= dest-namespace=argo-events dest-server="https://kubernetes.default.svc" diff_ms=2 fields.level=1 git_ms=6 health_ms=0 live_ms=1 settings_ms=0 sync_ms=0 time_ms=20
time="2022-04-06T10:44:46Z" level=info msg="Refreshing app status (controller refresh requested), level (1)" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="Comparing app state (cluster: https://kubernetes.default.svc, namespace: argo-events)" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="getRepoObjs stats" application=argo-events-ci build_options_ms=0 helm_ms=0 plugins_ms=0 repo_ms=0 time_ms=5 unmarshal_ms=5 version_ms=0
time="2022-04-06T10:44:46Z" level=info msg="Skipping auto-sync: most recent sync already to 453b2e55216b2de40fd95d008f96f24a654a3691" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="No status changes. Skipping patch" application=argo-events-ci
time="2022-04-06T10:44:46Z" level=info msg="Reconciliation completed" application=argo-events-ci dedup_ms=0 dest-name= dest-namespace=argo-events dest-server="https://kubernetes.default.svc" diff_ms=2 fields.level=1 git_ms=5 health_ms=0 live_ms=1 settings_ms=0 sync_ms=0 time_ms=20
```

**Comment by rishabh625** [comment]
can u check in logs what exactly is triggering excess reconcile

as log included here and released in v2.3.0, can't see in added logs
https://github.com/argoproj/argo-cd/pull/8192/files

**Comment by mikutas** [comment]
In our environment ExternalSecret/ClusterExternalSecret cause this issue

**Comment by ku524** [comment]
@mikutas Could you explain your case, please?
I experienced same issue nowadays, and coincidentally, we are also using external-secret

**Comment by iamKunal** [comment]
I faced a similar issue recently on my local cluster.

In my case I was using External Secrets too, but I was also using it for syncing Argo's username, password (using ES's templating features):
https://github.com/iamKunal/home-k8s-gitops-mirror/blob/d2d9f1f8e6752cc487a21e0d00be6be3330963a1/apps/argocd/templates/argocd-password-external-secret.yaml#L17-L18

Since ArgoCD creates the `server.secretKey` on startup I changed the creation Policy to merge so that there are no conflicts. But this led to the generated secret not being owned by the External Secret.

It led to some sort of weird loop wherein the External Secret update led to Secret Update which triggered reconciliation which led to External Secret update again.


I had to disable the secret generation by argo and disable the merging, but as others mentioned `server.secretKey` had to be set to an empty string. After an argoCD restart and it seems to be working fine once the secretKey is regenerated, but it does lead to a weird scenario with two keys in the secret having different values:

![image](https://user-images.githubusercontent.com/12411673/225581797-568d85a1-70e7-46a9-bf02-2dcbf64526b0.png)


Everything seems to be working fine now even after a couple of external secret refreshes.

On a separate note, it would be good to have https://github.com/argoproj/argo-cd/issues/10393

For reference, these are the changes: https://github.com/iamKunal/home-k8s-gitops-mirror/commit/da2f3e0b73bad758f52e8f2fdce51b124ac949da
